File: go/chapter04/question4.9/question4_9.go
// Author: Dinesh Appavoo (dineshappavoo) <dinesha.cit@gmail.com>

/*Algorithm FIND_PATH_EQUALS_SUM(t, path, sum, level): 
1.Get the input tree t, path array/hashmap datastructure, sum , level
2.if t==nil then return 
3.path[level]=t.Value
4.for i from level to 0 
	val+=path[i]
	if val == sum then print the path
5. Call left and right sub tree
	 FIND_PATH_EQUALS_SUM(t.Left, path, sum, level) 
	 FIND_PATH_EQUALS_SUM(t.Right, path, sum, level):
*/
package main

import (
	"fmt"
	"go/chapter04/binarytree"
)

func main() {

	inArr := []int{4, 5, 7, 8, 9, 10, 12, 14, 15, 17, 19, 20}
	t1 := binarytree.NewMinimalHeightBST(inArr, 0, len(inArr)-1)

	findAndPrintPath(t1)
}

func findAndPrintPath(t *binarytree.Tree) {
	height := binarytree.Height(t)
	var path = make([]int, int(height))
	findPathEqualsSum(t, path, 15, 0)
}

func findPathEqualsSum(t *binarytree.Tree, path []int, sum int, level int) {
	if t == nil {
		return
	}
	path[level] = t.Value
	val := 0
	for i := level; i >= 0; i-- {
		val += path[i]
		if val == sum {
			printPath(path, i, level)
		}
	}
	findPathEqualsSum(t.Left, path, sum, level+1)
	findPathEqualsSum(t.Right, path, sum, level+1)
}

func printPath(path []int, start int, end int) {
	fmt.Print("[ ")
	for i := start; i <= end; i++ {
		fmt.Print(path[i], " ")
	}
	fmt.Print("]\n")
}

File: c++/Chapter 4/Question4_9.cpp
#include<iostream>
using namespace std;

struct node{
    int data;
    node *right, *left;
};

void createMinimalBst(node* &root, int arr[], int start, int end){
    if(start>end)
        return;
    if(root==NULL){
        node *ptr = new node;
        int ind = start + (end-start)/2;
        ptr->data = arr[ind];
        ptr->left = NULL;
        ptr->right = NULL;
        root = ptr;
        createMinimalBst(root->left, arr, start, ind-1);
        createMinimalBst(root->right, arr, ind+1, end);
    }
}

void printIfSum(int* arr, int sum, int len){
    for(int i=0; i<len; i++){
        int s=0;
        for(int j=i; j<len; j++){
            s += arr[j];
            if(s == sum){
                for(int k=i; k<=j; k++)
                    cout<<arr[k]<<" ";
                cout<<'\n';
            }
        }
    }
}

void print(node *root, int sum, int *path, int level){
    if(root==NULL){
        return;
    }
    path[level] = root->data;
    int s = 0;
    for(int i=level; i>=0; i--){
        s += path[i];
        if(s==sum){
            for(int j=level; j>=i; j--)
                cout<<path[j]<<" ";
            cout<<'\n';
        }
    }
    if(root->left)
        print(root->left, sum, path, level+1);
    if(root->right)
        print(root->right, sum, path, level+1);
}

int main(){
    int arr[] = {1, 2, 3, 4, 5, 6, 7, 8, 9};
    node *root;
    root = NULL;
    createMinimalBst(root, arr, 0, 8);
    int path[100];
    print(root, 7, path, 0);
}
